rfcs/20201105-attribute-num-threads/README.md

```diff
@@ -0,0 +1,271 @@
+
+The options below are listed in the order of potentially higher changes in the
+library to support the option.
+
+Additional options are welcome to the list.
+
+### Option 1: new API in `dnnl::threadpool_interop` namespace
+
+This option proposes to introduce new API that limits the number of threads used
+for problem decomposition.
+
+<details>
+<summary>C API</summary>
+
+~~~cpp
+dnnl_status_t DNNL_API dnnl_threadpool_interop_primitive_attr_create(dnnl_primitive_attr_t *attr, uint32_t num_threads);
+~~~
+
+</details>
+
+<details>
+<summary>C++ API</summary>
+
+~~~cpp
+namespace dnnl {
+namespace threadpool_interop {
+
+inline dnnl::primitive_attr make_primitive_attr(uint32_t num_threads = DNNL_MAX_THREADS) {
+    dnnl_primitive_attr_t c_attr;
+    dnnl::error::wrap_c_api(dnnl_threadpool_interop_primitive_attr_create(&c_attr, num_threads));
+}
+
+} // namespace threadpool_interop
+} // namespace dnnl
+~~~
+
+</details>
+
+The `num_threads` value in the primitive attributes will indicate the upper
+limit for the number of threads used for decomposition in the primitive
+(descriptor). The default wildcard value `DNNL_MAX_THREADS` indicates that
+`dnnl_get_max_threads()` will be used internally as an upper limit.
+
+By assigning `0` in the API, the user will instruct the library that no
+threadpool will be used to execute the corresponding primitive, therefore no
+decomposition will be performed and the execution will be performed by the
+calling thread.
+
+By assigning `1` or higher value in the API, the user will instruct the library
+the upper limit for the number of threads for decomposition, and, during
+execution, the parallel sections will be executed by the threads in the
+threadpool.
+
+In the context of primitive cache, the number of threads passed by primitive
+attributes will be used to construct the cache key. In means that two primitives
+created using different `num_threads` will both be stored in the cache as
+separate implementations.
+
+Since the implementation of this RFC can target oneDNN v2.1 at the least, it is
+proposed to introduce the new API into `dnnl_threadpool.hpp`, where Threadpool-specific API is introduced.
+
+Pros:
+- Ability to have different problem decompositions for different primitives;
+- Flexibility with respect to user cost model and Threadpool implementation;
+- Minimal changes required in the library;
+
+Cons:
+- Upper limit for number of threads not applicable to other threading runtimes,
+  may cause some inconsistencies in the library code.
+
+### Option 2: new primitive attributes for all CPU threading runtimes
+
+In order to maintain consistency between threading runtimes, this option
+proposes to introduce a generic primitive attribute, that instructs the library
+the upper limit for thread count during primitive (descriptor) creation.
+
+<details>
+<summary>C API</summary>
+
+~~~cpp
+dnnl_primitive_attr_get_num_threads(dnnl_primitive_attr_t attr, uint32_t *num_threads);
```

Comment:
Can `num_threads` be bigger than number of real cores available on CPU?
Maybe rename `num_threads` to `max_num_threads` to better indicate the purpose of this attribute?

Reply:
I agree with your suggestions, Igor. Will update the RFC accordingly.

Reply:
I have updated the primitive attribute API for option 2 to emphasize its meaning.

Reply:
Regarding your first question, the answer is "it depends" and "it's debatable".

For the first option, since we are limited to the Threadpool runtime, the `max_num_threads` argument can be bigger than the number of available CPU cores in the system. In this case, the `max_num_threads` decomposition will be balanced among threads, and each thread will execute a chunk. To maintain this generality, perhaps the `max_num_threads` should be renamed to `max_decomposition_size`, however, as it is described in the RFC, having a decomposition higher than the number of threads is not optimal.

For the second option, in order to reduce potential oversubscription for OpenMP (and I guess TBB), `max_num_threads` has to be less or equal to the number of cores in the system (the implementation could include some checks for this).